src/tools/metrics/recorder.rs

```diff
@@ -0,0 +1,214 @@
+//! Metrics recording types and utilities.
+
+use std::collections::HashMap;
+
+use histogram::Histogram;
+use metrics::Key;
+use metrics_util::{
+    debugging::{DebugValue, DebuggingRecorder, Snapshotter},
+    CompositeKey, MetricKind,
+};
+
+// This is a false positive, `CompositeKey` does not depend on any interior mutable
+// types for its hashing implementation. Therefore, it is safe to use in our context.
+#[allow(clippy::mutable_key_type)]
+pub struct Snapshot(HashMap<CompositeKey, MetricVal>);
+
+impl Snapshot {
+    pub fn get_counter(&self, metric: &'static str) -> u64 {
```

Comment:
Can we pass `Key` as the second arg instead?

Reply:
seems like most of the API uses the metric name (`&str`), maybe we can keep it for consistency reasons

from below:
```rust
    fn can_get_counter_value() {
        let metrics = TestMetrics::default();
        let counter = register_counter!(METRIC_NAME);

        counter.increment(COUNTER_INC);

        let snapshot = metrics.take_snapshot();

        assert_eq!(snapshot.get_counter(METRIC_NAME), COUNTER_INC);
```